Copy old PlaneData attributes to new PlaneData

psavery opened the pull request:
Because PlaneData objects have immutable HKLs, when something happens that causes the HKLs to change (such as a change in dmin or sgnum) , a new PlaneData object must be instantiated with the new HKLs.

This PR does two main things:
1. Skips creating a new PlaneData object if the HKLs match. So small dmin changes, for instance, won't create a new PlaneData object.
2. Copies over old PlaneData attributes to the new PlaneData, including exclusions.

The exclusions are copied over by creating a mapping between the old HKLs and the new HKLs, and using this to set the new exclusions based on the old ones.

You can see the exclusions persisting here in hexrdgui:

![example gif](https://user-images.githubusercontent.com/9558430/110664434-00e03e80-818d-11eb-927e-01bdeab12a3a.gif)

Notably as well, other attributes such as the tThWidth and tThMax get preserved when a new PlaneData object is created also.

This, along with hexrd/hexrdgui#791, will resolve hexrd/hexrdgui#774

pep8speaks commented:
Hello @psavery! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`hexrd/material.py`](https://github.com/HEXRD/hexrd/blob/75853e0447cbbd6cde5edb0a5841dfda8382b040/hexrd/material.py):

> [Line 221:22](https://github.com/HEXRD/hexrd/blob/75853e0447cbbd6cde5edb0a5841dfda8382b040/hexrd/material.py#L221): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 221:22](https://github.com/HEXRD/hexrd/blob/75853e0447cbbd6cde5edb0a5841dfda8382b040/hexrd/material.py#L221): [E701](https://duckduckgo.com/?q=pep8%20E701) multiple statements on one line (colon)
> [Line 221:23](https://github.com/HEXRD/hexrd/blob/75853e0447cbbd6cde5edb0a5841dfda8382b040/hexrd/material.py#L221): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

psavery commented:
@saransh13 I want to make sure I understand which attributes on the `Material` object changing can change which HKLs come out of `self.unitcell.getHKLs()`. Can you confirm that I'm right about these? ("yes" if it changes which HKLs come out, "no" for does not).

1. dmin: yes
2. sgnum: yes
3. beamEnergy: no
4. latticeParameters: no
5. atominfo: no

Actually, now that I think about it, because 3 and 4 can change the dmin of the different HKLs, they could potentially change which HKLs come out, because they might move some HKLs above/below the dmin. Is that right? If that's the case, all of those attributes change the HKLs except atominfo...

saransh13 commented:
Hey Patrick,

You’re right. All of them except the atominfo changes the hkl list.

On Wed, Mar 10, 2021 at 8:55 AM Patrick Avery ***@***.***>
wrote:

> @saransh13 <https://github.com/saransh13> I want to make sure I
> understand which attributes on the Material object changing can change
> which HKLs come out of self.unitcell.getHKLs(). Can you confirm that I'm
> right about these? ("yes" if it changes which HKLs come out, "no" for does
> not).
>
>    1. dmin: yes
>    2. sgnum: yes
>    3. beamEnergy: no
>    4. latticeParameters: no
>    5. atominfo: no
>
> Actually, now that I think about it, because 3 and 4 can change the dmin
> of the different HKLs, they could potentially change which HKLs come out,
> because they might move some HKLs above/below the dmin. Is that right? If
> that's the case, all of those attributes change the HKLs except atominfo...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/HEXRD/hexrd/pull/204#issuecomment-795732023>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADYZ2U3SNZAF4NKPURPFIFLTC6I6VANCNFSM4Y6NJERQ>
> .
>

joelvbernier commented:
Slight correction to the above; changes in latticeParameters *can* (as in will generally) change the generated hkls up to dmin.

psavery commented:
@joelvbernier Sounds good, I'll fix that soon

psavery commented:
@saransh13 which of those 5 properties affect the structure factor? All of them?

saransh13 commented:
Hey Patrick,

So the beamenergy, lattice parameter and atominfo will affect the structure
factor. Since sgnum affects some of them, it will indirectly change
structure factors too.
It is independent of dmin.

On Thu, Mar 11, 2021 at 11:34 AM Patrick Avery ***@***.***>
wrote:

> @saransh13 <https://github.com/saransh13> which of those 5 properties
> affect the structure factor? All of them?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/HEXRD/hexrd/pull/204#issuecomment-796992608>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADYZ2U3STZG55PIRN7GXLQDTDEEKTANCNFSM4Y6NJERQ>
> .
>

psavery commented:
@saransh13 Thanks saransh! I have updated the code to make sure the structure factor gets updated for all of these changes.

This PR is now ready.

PS: I removed the comparison of hkls to skip generating new PData, since generating new PData is not that expensive. It simplifies a little bit of the logic.

saransh13 commented on hexrd/material.py:
```diff
@@ -775,9 +785,9 @@ def _set_atominfo(self, v):
+            raise ValueError("Improper syntax, array must be n x 4")
 
-        return
+        self.update_structure_factor()
```
Not related to your commit, but I noticed that the atominfo is not getting updated in the unitcell class, which is what actually calculates the structure factors.
I'll push a fix for this soon.

saransh13 reviewed (commented):
Looks good to me overall.

saransh13 commented on hexrd/material.py:
```diff
@@ -230,6 +230,7 @@ def _newUnitcell(self):
     def _hkls_changed(self):
         # Call this when something happens that changes the hkls...
         self._newPdata()
+        self.update_structure_factor()
```
Great! LGTM.

joelvbernier commented:
Ah crap -- @saransh13 looks like the more recent changes in #218 caused a conflict.  Looks like an easy resolution; let me or @psavery know if you need help.

joelvbernier commented:
@saransh13 and @psavery -- I did my best to resolve the merge conflict with this PR; please double-check.  If it looks good, LGTM!

joelvbernier commented:
@psavery, can you give a quick check as well?

saransh13 reviewed (commented):
LGTM
